### ISSUE TITLE
Multiple loss functions and multiple optimisers

### ISSUE THREAD

**Issue by raymondchua** [deepmind/optax#350]
Hi, is there a way to modify [this](https://github.com/deepmind/optax/issues/59#issuecomment-800747919) for situations with multiple loss functions? For example, I have two loss functions and I partition the parameters of my network in two sets (eg. set 1 and set 2). For update step that is odd (step 1,3,5..), I would like to update parameters set 1 with loss function 1. For update step that even (step 2,4, 6...), I would like to update parameters set 2 with loss function 2.

**Comment by n2cholas** [comment]
You can write a relatively simple `alternate_tx` wrapper that will choose a different optimizer based on the step. 

```py
from functools import partial
from typing import NamedTuple, Any

import jax
import jax.numpy as jnp

import optax

class AlternateTxState(NamedTuple):
    step: jnp.ndarray
    tx1_state: Any
    tx2_state: Any

def alternate_tx(tx1, tx2):
    def init_fn(params):
        return AlternateTxState(
            step=jnp.zeros([], dtype=jnp.int32),
            tx1_state=tx1.init(params),
            tx2_state=tx2.init(params),
        )
    
    def _update_tx1(updates, state, params=None):
        new_updates, new_state = tx1.update(updates, state.tx1_state, params=None)
        return new_updates, state._replace(step=state.step+1, tx1_state=new_state)

    def _update_tx2(updates, state, params=None):
        new_updates, new_state = tx2.update(updates, state.tx2_state, params=None)
        return new_updates, state._replace(step=state.step+1, tx2_state=new_state)

    def update_fn(updates, state, params=None):
        return jax.lax.cond(
            state.step % 2 == 0,
            _update_tx1,
            _update_tx2,
            updates, state, params
        )

    return optax.GradientTransformation(init_fn, update_fn)
```

Usage:

```py
import haiku as hk

def forward_fn(x):
  return hk.Sequential([
    hk.Linear(10),
    hk.Linear(10),
  ])(x)
forward = hk.transform(forward_fn)

params = forward.init(jax.random.PRNGKey(0), jnp.ones([1, 1]))
param_label_fn = partial(hk.data_structures.map, lambda mname, name, val: name)

optimizer = alternate_tx(
    # only update kernels, and set the update value for biases to 0 so they are untouched
    optax.multi_transform({'w': optax.sgd(learning_rate=0.01), 'b': optax.set_to_zero()}, param_label_fn),
    # same as above, except only update biases instead of kernels
    optax.multi_transform({'b': optax.sgd(learning_rate= 0.2), 'w': optax.set_to_zero()}, param_label_fn)) 
```

Here's some code demonstrating what happens:

```py
state = optimizer.init(params)
gradients = jax.tree_map(jnp.ones_like, params)

updates_1, state_1 = optimizer.update(gradients, state)
updates_2, state_2 = optimizer.update(gradients, state_1)
```

In `updates_1`, you'll see that only kernels have update values while biases are zero'd. In `updates_2`, you'll see the opposite.

Does this make sense? This can be extended to support an arbitrary number of inner transformations with some work.

I realize it has been almost a month since you asked the question, so please share if you've come up with an alternative solution :).

**Comment by raymondchua** [comment]
@n2cholas , I think they can take place in the same step as long as representation parameters are updated before the prediction's parameters. I imagine my code snippet does that?

**Comment by n2cholas** [comment]
If that's the case, then yes your code snippet looks fine, and you should ignore mine (mine assumes you want to compute the loss with the old parameters, compute gradients, update representation params, evaluate loss with new representation parameters and old prediction parameters, compute gradients with this loss, update prediction parameters, then repeat).

**Comment by raymondchua** [comment]
Thank you so much for your help @n2cholas :)
